### ISSUE TITLE
Doors bugs

### ISSUE THREAD

**Issue by ShadowNinja** [minetest/minetest_game#385]
* [ ] Nodes are not checked for ignore, leading to posibly replacing solid nodes.
  * [x] `buildable_to` of `pointed_thing.under` is not respected.

This is just from a cursory examination.  The mod is also full of odd variable names and ugliness, and could stand a thourough cleanup.

**Comment by MT-Modder** [comment]
@PilzAdam When you clean up or rewrite this mod could you add doors to a global table like this:

``` doors.registered_doors ```

So mods like mesecons can easily rewrite the door definitions.
I'm not sure how easy this would be considering each door is two nodes.

**Comment by C1ffisme** [comment]
BUMP!

**Comment by paramat** [comment]
@PilzAdam
> I could clean this mod up after #376 is merged.

**Comment by PilzAdam** [comment]
Sorry, I'm currently busy.

**Comment by MT-Modder** [comment]
The doors mod needs to be reanalyzed for bugs because most if not all of the previous code has been changed/removed after https://github.com/minetest/minetest_game/commit/f600a9f645af40d22c8eb7c17aff89507b71816e. I'm pretty sure the same bugs remain.

**Comment by paramat** [comment]
@sofar please could you check the new code addresses the issues here? (if ShadowNinja's requests are reasonable).

**Comment by sofar** [comment]
@paramat yeah, totally, worth the effort as well.

**Comment by sofar** [comment]
@paramat please assign to me?

**Comment by MT-Modder** [comment]
The second bug listed is certainly still there, I tested.

**Comment by sofar** [comment]
@MT-Modder yup, the new doors code is an improvement, but certainly not a finished cleanup version.

**Comment by sofar** [comment]
Item 2 is fixed by #850  - @mt-modder perhaps can test this?

**Comment by sofar** [comment]
Item 3 is no longer present. metadata is set in the ```on_place()``` code without further tricks.

**Comment by MT-Modder** [comment]
Tested!

Works as it should but I did encounter another bug. You can't use doors while holding a door. It just places another door next to it.

**Comment by sofar** [comment]
> You can't use doors while holding a door. It just places another door next to it.

Believe it or not but that used to actually work. I'll see if I can restore that functionality.

**Comment by MT-Modder** [comment]
I thought it seemed odd.

**Comment by sofar** [comment]
#850 also has a second patch that fixes the right-click behavior.

**Comment by MT-Modder** [comment]
Placing a door isn't logged in console/debug. Digging is though. I'll test again.

**Comment by MT-Modder** [comment]
Tested and works.

**Comment by sofar** [comment]
Not having placement logs should be a separate bug, there's likely dozens of node handlers that are omitting logging functions.

**Comment by MT-Modder** [comment]
Yeah it's probably more of an engine bug.

**Comment by paramat** [comment]
I can't assign you as you are not core dev.
Removed item 3 from first comment.

**Comment by sofar** [comment]
I've spent quite some time attempting to hit ```nil``` conditions at 30000 elevation and higher, but couldn't make it crash. I don't think it's a critical issue, if someone can show me a backtrace of an actual bug, it'll likely be easy to fix anyway. With the rest of the fixes pending, i think this can be closed once the latest door update gets merged.

**Comment by paramat** [comment]
#850 is merged.

**Comment by paramat** [comment]
Is the first point, checking for ignore, an issue?

**Comment by MT-Modder** [comment]
As @ShadowNinja said:
>@MT-Modder Unloaded nodes show up as ignore, which is buildable_to. The real node is overwritten as soon as the block is loaded.

So non-buildable_to blocks will get replaced when they shouldn't. It's rare but possible.

**Comment by paramat** [comment]
Mapgen doesn't replace pre-existing nodes, so if a door is placed during mapgen in ignore it should survive. Ignore won't be present when placing manually. Closing.

**Comment by MT-Modder** [comment]
Right it wouldn't be the mapgen doing the replacing it would be a mod. And the door wouldn't get replaced, the node that would have been generated gets replaced when it shouldn't have. I ran into this problem with a mod I made. It kept trying to replace ignore with air and throwing errors like crazy. Again it's rare but possible.

And in the rare case where a part of the map is ungenerated (taking a while) and a player places a door on what would be dirt it get's replaced by the door when it shouldn't. I've never actually seen this but that's the theory anyway. A lot of mapgen optimizations have been made in the past year making running into ignore only possible when using fast_mode, for me anyway.
